subsys/caf/modules/sensor_sampler.c

```diff
@@ -0,0 +1,201 @@
+static K_THREAD_STACK_DEFINE(sample_thread_stack, SAMPLE_THREAD_STACK_SIZE);
+static struct k_thread sample_thread;
+
+
+static void report_error(void)
+{
+	state = STATE_ERROR;
+	module_set_state(MODULE_STATE_ERROR);
+}
+
+static void send_sensor_event(const char *descr, struct sensor_value *data, size_t data_cnt)
+{
+	struct sensor_event *event = new_sensor_event(sizeof(float) * data_cnt);
+	float *data_ptr = sensor_event_get_data_ptr(event);
+
+	event->descr = descr;
+
+	__ASSERT_NO_MSG(sensor_event_get_data_cnt(event) == data_cnt);
+	for (size_t i = 0; i < data_cnt; i++) {
+		data_ptr[i] = sensor_value_to_double(&data[i]);
+	}
+
+	EVENT_SUBMIT(event);
+}
+
+static int sample_sensor(const struct device *dev, const struct sensor_config *sc)
+{
+	size_t data_cnt = 0;
+
+	for (size_t i = 0; i < sc->chan_cnt; i++) {
+		data_cnt += sc->chans[i].data_cnt;
+	}
+
+	int err = 0;
+	size_t data_idx = 0;
+	struct sensor_value data[data_cnt];
+
+	for (size_t i = 0; !err && (i < sc->chan_cnt); i++) {
+		const struct sampled_channel *sampled_chan = &sc->chans[i];
+
+		err = sensor_sample_fetch_chan(dev, sampled_chan->chan);
+		if (!err) {
+			err = sensor_channel_get(dev, sampled_chan->chan, &data[data_idx]);
+			data_idx += sampled_chan->data_cnt;
+		}
+	}
+
+	if (err) {
+		LOG_ERR("Sensor sampling error (err %d)", err);
+	} else {
+		send_sensor_event(sc->event_descr, data, ARRAY_SIZE(data));
+	}
+
+	return err;
+}
+
+static int sample_sensors(int64_t *next_timeout)
+{
+	int err = 0;
+	int64_t cur_uptime = k_uptime_get();
+
+	*next_timeout = INT64_MAX;
+
+	for (size_t i = 0; !err && (i < ARRAY_SIZE(sensor_data)); i++) {
+		struct sensor_data *sd = &sensor_data[i];
+		const struct sensor_config *sc = &sensor_configs[i];
+
+		if (sd->sample_timeout <= cur_uptime) {
+			err = sample_sensor(sd->dev, sc);
+
+			while (sd->sample_timeout <= cur_uptime) {
+				sd->sample_timeout += sd->sampling_period;
+			}
+		}
+
+		if (*next_timeout > sd->sample_timeout) {
+			*next_timeout = sd->sample_timeout;
+		}
+	}
+
+	return err;
+}
+
+static int sensor_init(void)
+{
+	int err = 0;
+	int64_t cur_uptime = k_uptime_get();
+
+	for (size_t i = 0; !err && (i < ARRAY_SIZE(sensor_data)); i++) {
+		struct sensor_data *sd = &sensor_data[i];
+		const struct sensor_config *sc = &sensor_configs[i];
+
+		sd->dev = device_get_binding(sc->dev_name);
+		if (!sd->dev) {
+			err = -ENXIO;
+			break;
+		}
+		sd->sampling_period = sc->sampling_period_ms;
+		sd->sample_timeout = cur_uptime + sc->sampling_period_ms;
+	}
+
+	if (!err) {
+		state = STATE_ACTIVE;
+		module_set_state(MODULE_STATE_READY);
+	} else {
+		report_error();
+	}
+
+	return err;
+}
+
+static void sample_thread_fn(void)
+{
+	int err = 0;
+	int64_t next_timeout;
+
+	err = sensor_init();
+
+	while (!err) {
+		err = sample_sensors(&next_timeout);
+
+		__ASSERT_NO_MSG(next_timeout != INT64_MAX);
+		int64_t sleep_ms = next_timeout - k_uptime_get();
+
+		if (sleep_ms < 0) {
+			sleep_ms = 0;
+		}
+		k_sleep(K_MSEC(sleep_ms));
```

Comment:
Small problem in our case but if we only had `k_sleep_until` in the kernel....

@pabigot Of topic to this review. Do you think `k_sleep_until(tick)` would be useful to add?

Reply:
AFAIK it should be possible to invoke `k_sleep` with an absolute `k_timeout_t`.  Would `k_sleep_until()` do something different?

Reply:
No, that would be only to sleep until absolute time.

Does k_sleep do it? The implementation would suggest this is only relative https://github.com/zephyrproject-rtos/zephyr/blob/master/kernel/sched.c#L1257

Reply:
`K_TIMEOUT_ABS_TICKS`?

Reply:
@MarekPieta if it works please use it

Reply:
> Does k_sleep do it? The implementation would suggest this is only relative https://github.com/zephyrproject-rtos/zephyr/blob/master/kernel/sched.c#L1257

That's unfortunate.  Absolute timeouts really aren't a first-class citizen.

> No, that would be only to sleep until absolute time.

Without searching for context I'm not sure what the problem is, or what `k_sleep_until(ticks)` would do.  By its name I assumed it was intended to wait until a specific absolute tick (time).

I've pushed for absolute timeouts because, given Zephyr's timer implementation, it's the closest we can get to scheduling an operation at an absolute time, or repeating operations at a fixed interval regardless of latency.   My response was to suggest something that I thought might accomplish that.  But apparently it won't.

Reply:
thanks for the comments